### PR TITLE
bug: missing preprocessing examples for automl edge models

### LINKED ISSUE #130: Preprocessing is missing in the current edge model samples, and might cause low performances
Preprocessing is missing in the current edge model samples, and might cause low performances.

### PR DESCRIPTION
**Pull request by genquan9**
Fixes #130

### CODE REVIEW

File: samples/vision_edge/edge_container_predict/automl_vision_edge_container_predict.py [hunk 1]
```diff
@@ -18,18 +18,45 @@
...
 import argparse
 # [START automl_vision_edge_container_predict]
 import base64
+import cv2
 import io
 import json
 
 import requests
 
+def preprocess_image(image_file_path, max_width, max_height):
```
Comment (telpirion) [inline samples/vision_edge/edge_container_predict/automl_vision_edge_container_predict.py:34]: This function is called `preprocess_image()` but it looks like the main thing it does is resize--is that correct? If so, I suggest renaming it.

File: samples/vision_edge/edge_container_predict/automl_vision_edge_container_predict.py [hunk 2]
```diff
@@ -18,18 +18,45 @@
...
+        image_file_path: Path to a local image for the prediction request.
+        max_width: The max width for preprocessed images. The max width is 640 
+            (1024) for AutoML ICN (IOD) edge models.
+        max_height: The max width for preprocessed images. The max height is  
+            480 (1024) for AutoML ICN (IOD) edge models.
+    Returns:
+        The preprocessed encoded image bytes.
+    """
+    encode_param = [int(cv2.IMWRITE_JPEG_QUALITY), 85]
+    im = cv2.imread(image_file_path)
```
Comment (telpirion) [inline samples/vision_edge/edge_container_predict/automl_vision_edge_container_predict.py:47]: Also, the reason for using the `cv2` library is never explained. We shouldn't add new dependencies to customer's solutions without some sort of reason why. Please add a comment that explains why we need to use the library here.

File: samples/vision_edge/edge_container_predict/automl_vision_edge_container_predict.py [hunk 3]
```diff
@@ -41,9 +68,11 @@ def container_predict(image_file_path, image_key, port_number=8501):
     Returns:
         The response of the prediction request.
     """
-
-    with io.open(image_file_path, 'rb') as image_file:
-        encoded_image = base64.b64encode(image_file.read()).decode('utf-8')
+    # AutoML edge models will preprocess the input image.
```
Comment (telpirion) [inline samples/vision_edge/edge_container_predict/automl_vision_edge_container_predict.py:71]: Check branding guidelines: "Edge" should be capitalized and should use "Vision" before it.

File: samples/vision_edge/edge_container_predict/automl_vision_edge_container_predict.py [hunk 4]
```diff
@@ -41,9 +68,11 @@ def container_predict(image_file_path, image_key, port_number=8501):
...
+    # The max width and height for AutoML ICN (IOD) edge models are 640*480
```
Comment (telpirion) [inline samples/vision_edge/edge_container_predict/automl_vision_edge_container_predict.py:72]: Remove abbreviations and use branded names: AutoML Vision Classification and AutoML Vision Image Object Detection.

File: samples/vision_edge/edge_container_predict/automl_vision_edge_container_predict.py [hunk 5]
```diff
@@ -41,9 +68,11 @@ def container_predict(image_file_path, image_key, port_number=8501):
...
+    # (1024*1024). The example here is for ICN edge models.
```
Comment (telpirion) [inline samples/vision_edge/edge_container_predict/automl_vision_edge_container_predict.py:73]: See previous comment.

**Review by telpirion — CHANGES_REQUESTED** [review]

**Comment by sirtorry** [comment]
OOO right now. looks good to me on glance.

File: samples/vision_edge/edge_container_predict/automl_vision_edge_container_predict.py [hunk 6]
```diff
@@ -18,18 +18,45 @@
...
 import json
 
 import requests
 
+def preprocess_image(image_file_path, max_width, max_height):
+    """Preprocesses input images for AutoML edge models.
+    
+    Args:
+        image_file_path: Path to a local image for the prediction request.
+        max_width: The max width for preprocessed images. The max width is 640
```
[the remark this answers is not in the document]
Comment (genquan9) [reply samples/vision_edge/edge_container_predict/automl_vision_edge_container_predict.py:39]: There is a space in the end. Removed.

File: samples/vision_edge/edge_container_predict/automl_vision_edge_container_predict.py (hunk 1 above)
Comment (genquan9) [reply samples/vision_edge/edge_container_predict/automl_vision_edge_container_predict.py:34]: The current implementation mainly did read, resize and encoded it as string.
Will preprocess be a better name?

File: samples/vision_edge/edge_container_predict/automl_vision_edge_container_predict.py (hunk 2 above)
Comment (genquan9) [reply samples/vision_edge/edge_container_predict/automl_vision_edge_container_predict.py:47]: Added comments.

File: samples/vision_edge/edge_container_predict/automl_vision_edge_container_predict.py (hunk 3 above)
Comment (genquan9) [reply samples/vision_edge/edge_container_predict/automl_vision_edge_container_predict.py:71]: done

File: samples/vision_edge/edge_container_predict/automl_vision_edge_container_predict.py (hunk 4 above)
Comment (genquan9) [reply samples/vision_edge/edge_container_predict/automl_vision_edge_container_predict.py:72]: done.

File: samples/vision_edge/edge_container_predict/automl_vision_edge_container_predict.py (hunk 5 above)
Comment (genquan9) [reply samples/vision_edge/edge_container_predict/automl_vision_edge_container_predict.py:73]: done

File: samples/vision_edge/edge_container_predict/automl_vision_edge_container_predict.py [hunk 7]
```diff
@@ -37,10 +37,10 @@ def preprocess_image(image_file_path, max_width, max_height):
     Args:
         image_file_path: Path to a local image for the prediction request.
         max_width: The max width for preprocessed images. The max width is 640
-            (1024) for AutoML Vision image classfiication (object detetion)
+            (1024) for AutoML Vision image classfication (object detetion)
```
Comment (telpirion) [inline samples/vision_edge/edge_container_predict/automl_vision_edge_container_predict.py:40]: NIT: I think that the branding calls for initial capitals: AutoML Vision Image Classification. See here:
https://cloud.google.com/vision/overview/docs#automl-vision

**Review by telpirion — APPROVED** [review]

File: samples/vision_edge/edge_container_predict/automl_vision_edge_container_predict.py (hunk 7 above)
Comment (telpirion) [reply samples/vision_edge/edge_container_predict/automl_vision_edge_container_predict.py:40]: Also, "detection" is spelled wrong and we present Image Detection and Image Classification as two separate objectives.

Comment (genquan9) [reply samples/vision_edge/edge_container_predict/automl_vision_edge_container_predict.py]: done. Thanks for the comments!

**Pull request merged**